### PR TITLE
added blog post for theory of mind

### PR DESCRIPTION
**Pull request by ojus1**

### CODE REVIEW

File: _posts/2022-05-19-theory-of-mind.md [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+1. The ability of the agent to recognize its own errors in perceiving (mis-hearing), i.e., discover its own false-beliefs and ask for clarifications. This also leads to a higher order reasoning capability of the agent.
+2. The ability of the agent to dynamically model its counterpart throughout the conversation and adjust its own behaviour inorder to maximize the success of the dialog. Dynamic response and prosody generation, turn-taking, barge-in handling, etc. are such examples.
+
+## Do Machines have a Theory of Mind?
+
+One of the important goals of AI is to blend in the lives of Humans and solve problems _with humans-in-the-loop_, achieving this requires modeling humans and other machines around the agent, similar to how we humans do.
+
+Some studies [3, 4] have shown that specially designed Multi-Agent Reinforcement Learning algorithms pass the Sally-Anne False-belief task. However, False-belief tasks have not been designed/testsed against chatbots. In this section, we test multiple Language Models (LM) against the Sally-Anne task, and check whether they pass the test or not.
```
Comment (piyushmakhija5) [inline _posts/2022-05-19-theory-of-mind.md:91]: Spelling mistake. `testsed` --> `tested`

File: _posts/2022-05-19-theory-of-mind.md [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+There are two possible ways to "solve" this problem:
+1. `Explicit`: Design the product in a way that inspires trust. Come up with the _best possible_ responses for all possible combination of dialog history and user states.
+2. `Implicit`: Design the product in a top-down fashion rather than a bottom-up. Many believe that optimizing components with their local objective (Word-error-rate for ASR, F1 Scores for Intent classifiers etc.) would lead to a higher resolution rate. In biological systems, the higher-order function (survival) dictates lower order function (communication, language). Learning to communicate better can not ensure survival on its own. However, learning to survive _may lead to a better ability to communicate_. In other words, optimize ML models against objective (resolution rate) in-addition to the local objective. This _will_ force the chatbot to behave in a way that inspires trust from the user and effectively learn to have theory of mind of the users.
+
+Clearly, the first method doesn't seem to be working well, and the second method has the clear advantage of being data-driven and scalable.
```
Comment (piyushmakhija5) [inline _posts/2022-05-19-theory-of-mind.md:138]: A person not intricately working in this domain, might not be able to come to this conclusion "clearly".

File: _posts/2022-05-19-theory-of-mind.md [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+The participating agent is told the following scenario:
+1. Sally and Anne are inside a room. 
+2. Sally has a basket with one marble inside it.
+3. Anne has an empty box with her.
+4. Sally leaves the room without her basket.
+5. Anne takes the marble out of Sally's basket and puts it in her own box.
+6. Sally comes back inside the room.
+
+Now, the participant is asked, "Where will Sally look for her marble?". If the participant replies with "the basket", this means that the participant is able to model the mental state of a fictional character Sally, and that she doesn't know that Anne took her marble. Children below the age of 3-4 answer with "the box", however, older children answer with "the basket". Some children with mental disabilities such as Down syndrome and Autism are unable to pass this test.
```
Comment (piyushmakhija5) [inline _posts/2022-05-19-theory-of-mind.md:68]: pass/fail is somewhat unclear here. The first time I read this para, I thought both answers could be correct is what we are going for.

**Review by piyushmakhija5 — COMMENTED** [review]
Added a few review comments

File: _posts/2022-05-19-theory-of-mind.md (hunk 3 above)
Comment (ojus1) [reply _posts/2022-05-19-theory-of-mind.md:68]: Sally was outside the room when Anne swapped the place of the marble. According to Sally, the marble should be in her basket. However, the child (the participant) knows that the marble has been swapped. Agents who don't have a theory of mind _cannot_ model the right state of mind for Sally, instead, they say Sally will look for her marble in Anne's box first.

Does this make sense now?

File: _posts/2022-05-19-theory-of-mind.md (hunk 2 above)
Comment (ojus1) [reply _posts/2022-05-19-theory-of-mind.md:138]: Agreed, but I've pointed out that the first method is how the industry does things right now, and how bad it is.

File: _posts/2022-05-19-theory-of-mind.md [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+| DistilGPT2              | 82M      | Fail      | Fail      |
+| GPT-Neo-125M            | 125M     | Fail      | Fail      |
+| GPT-Neo-1.3B            | 1.3B     | Fail      | Fail      |
+| GPT-2                   | 1.5B     | Pass      | Pass      |
+| GPT-Neo-2.7B            | 2.7B     | Pass      | Pass      |
+| GPT-J-6B                | 6B       | Pass      | Pass      |
+
+The largest three of the models pass both the tests. This suggests that scale might help LMs achieve some basic reasoning capabilities. This result is not surprising, since larger LMs usually do better in reasoning benchmarks.
+
+P.S. The most entertaining response award goes to DistilGPT2 for `"I don't give a fig about the box"` for the Sally-Anne task. This is not made up, I swear!
```
Comment (swarajdalmia) [inline _posts/2022-05-19-theory-of-mind.md:119]: lol, this is funny.

File: _posts/2022-05-19-theory-of-mind.md [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+There are two possible ways to "solve" this problem:
+1. `Explicit`: Design the product in a way that inspires trust. Come up with the _best possible_ responses for all possible combination of dialog history and user states.
+2. `Implicit`: Design the product in a top-down fashion rather than a bottom-up. Many believe that optimizing components with their local objective (Word-error-rate for ASR, F1 Scores for Intent classifiers etc.) would lead to a higher resolution rate. In biological systems, the higher-order function (survival) dictates lower order function (communication, language). Learning to communicate better can not ensure survival on its own. However, learning to survive _may lead to a better ability to communicate_. In other words, optimize ML models against objective (resolution rate) in-addition to the local objective. This _will_ force the chatbot to behave in a way that inspires trust from the user and effectively learn to have theory of mind of the users.
+
+Clearly, the first method doesn't seem to be working well, and the second method has the clear advantage of being data-driven and scalable.
+
+# References
```
Comment (swarajdalmia) [inline _posts/2022-05-19-theory-of-mind.md:140]: These are not rendered properly in the markdown. Will have to add spacing.

**Pull request merged**